File: common/src/main/java/io/github/mqzn/commands/base/tree/SubCommandArgumentWrapper.java
package io.github.mqzn.commands.base.tree;

import io.github.mqzn.commands.arguments.Argument;
import io.github.mqzn.commands.base.syntax.SubCommandSyntax;
import org.jetbrains.annotations.NotNull;

import java.util.LinkedList;


public final class SubCommandArgumentWrapper<S> {
	
	@NotNull
	private final CommandTree<S> tree;
	
	@NotNull
	private final SubCommandSyntax<S> subCommand;
	
	@NotNull
	private final LinkedList<Argument<?>> totalArguments = new LinkedList<>();
	
	private SubCommandArgumentWrapper(@NotNull CommandTree<S> tree,
	                                  @NotNull SubCommandSyntax<S> subCommand) {
		this.tree = tree;
		this.subCommand = subCommand;
		collectArguments();
	}
	
	
	public static <S> SubCommandArgumentWrapper<S> wrap(@NotNull CommandTree<S> tree,
	                                                    @NotNull SubCommandSyntax<S> subCommand) {
		
		return new SubCommandArgumentWrapper<>(tree, subCommand);
	}
	
	private void collectArguments() {
		
		this.totalArguments.add(
			Argument.literal(subCommand.getName())
				.aliases(subCommand.getAliases().getArray())
		);
		
		this.totalArguments.addAll(subCommand.getArguments());
		
		String parent = subCommand.getParent();
		
		while (parent != null) {
			
			SubCommandSyntax<S> parentSubCommand = tree.getSubCommand(parent);
			if (parentSubCommand == null) {
				break;
			}
			
			
			for (var parentArg : parentSubCommand.getArguments()) {
				this.totalArguments.addFirst(parentArg);
			}
			
			this.totalArguments.addFirst(
				Argument.literal(parentSubCommand.getName())
					.aliases(parentSubCommand.getAliases().getArray())
			);
			
			parent = parentSubCommand.getParent();
		}
		
	}
	
	
	public SubCommandSyntax<S> get() {
		return subCommand;
	}
	
	public LinkedList<Argument<?>> parentalArguments() {
		return totalArguments;
	}
	
}

File: common/src/main/java/io/github/mqzn/commands/base/syntax/SubCommandArgumentWrapper.kt
package io.github.mqzn.commands.base.syntax

import io.github.mqzn.commands.arguments.Argument
import java.util.*

class SubCommandArgumentWrapper<S> private constructor(
    private val tree: CommandTree<S>,
    private val subCommand: SubCommandSyntax<S>
) {
    private val totalArguments = LinkedList<Argument<*>>()

    init {
        collectArguments()
    }

    private fun collectArguments() {
        totalArguments.add(
            Argument.literal(subCommand.name)
                .aliases(*subCommand.aliases.array)
        )
        totalArguments.addAll(subCommand.arguments)
        var parent = subCommand.parent
        while (parent != null) {
            val parentSubCommand = tree.getSubCommand(parent) ?: break
            for (parentArg in parentSubCommand.arguments) {
                totalArguments.addFirst(parentArg)
            }
            totalArguments.addFirst(
                Argument.literal(parentSubCommand.name)
                    .aliases(*parentSubCommand.aliases.array)
            )
            parent = parentSubCommand.parent
        }
    }

    fun get(): SubCommandSyntax<S> {
        return subCommand
    }

    fun parentalArguments(): LinkedList<Argument<*>> {
        return totalArguments
    }

    companion object {
        fun <S> wrap(
            tree: CommandTree<S>,
            subCommand: SubCommandSyntax<S>
        ): SubCommandArgumentWrapper<S> {
            return SubCommandArgumentWrapper(tree, subCommand)
        }
    }
}